### ISSUE TITLE
Scheduled notification doesn't work properly in iOS when called from didChangeAppLifecycleState

### ISSUE THREAD

**Issue by misici234** [MaikuB/flutter_local_notifications#1356]
**Describe the bug**
Scheduled notification doesn't work properly in iOS release build only when it's called inside didChangeAppLifecycleState just before the app goes to the background.

**To Reproduce**
- I used 
  flutter_local_notifications: ^9.0.1
in the pubspec.yaml

- Doctor summary (to see all details, run flutter doctor -v):
[✓] Flutter (Channel stable, 2.2.3, on macOS 11.6 20G165 darwin-x64, locale en-CA)
[✓] Android toolchain - develop for Android devices (Android SDK version 30.0.2)
[✓] Xcode - develop for iOS and macOS
[✓] Chrome - develop for the web
[✓] Android Studio (version 4.2)
[✓] Connected device (3 available)

- iOS version 15.0.1 but I noticed this problem in ver 14.4 too

1. Make the code changes described in the Sample section
2. Run "flutter build ipa" to generate Runner.xcarchive file
3. Open above archive file in Xcode and produce IPA file
4. Install IPA file on you iPhone
5. run the app and make sure you get notification from "Schedule notification to appear in 5 sec..."
6. Push the app to the background

**Expected behavior**
I expect to get "background title" notification in step 6. after about 6 to 10 seconds but I don't get it at all. Actually, I get it after I resume the app from the background which can happen 1 min, 1 hour or any time but this is too late for a reminder and it's not how it is supposed to work.
As soon as I activate the app and tap "Check pending notifications" I get 1 pending and after a few seconds the notifications is displayed with a default iOS sound. Why does this happen and how can it be resolved? This is very big issue for my company. Interestingly, the same scenario works correctly in Android. It also works correctly in iOS when you run the app in debug mode on iPhone. Only iOS build has this problem.

**Sample code to reproduce the problem**
Use the example app to reproduce this problem but make the following changes in main.dart
```
  @override
  void didChangeAppLifecycleState(AppLifecycleState state) {
    super.didChangeAppLifecycleState(state);

    final DateTime alcsChangedAt = DateTime.now().toUtc();

    switch (state) {
      case AppLifecycleState.inactive:
        print('App went to the foreground inactive state at $alcsChangedAt');
        break;

      case AppLifecycleState.paused:
        print('App went to background at $alcsChangedAt');
        Future<int>.value(1).then((int value) async {
          await Future.delayed(const Duration(seconds: 1)); // this is the most important to reproduce this problem
          // Above Future delay is a web API call, for example, to get upcoming appointments
          await _zonedScheduleNotification(title: 'background title', body: 'background body');
        });
        break;

      case AppLifecycleState.resumed:
        print('App resumed from background at $alcsChangedAt');
    }
  }
```
Do everything else so that didChangeAppLifecycleState can work properly like 
class _HomePageState extends State<HomePage> with WidgetsBindingObserver {
and so on.
Modify _zonedScheduleNotification as below (this is not so important but just to make testing cleaner and more obvious)

```
  Future<void> _zonedScheduleNotification({
    String title = 'scheduled title',
    String body = 'scheduled body',
  }) async {
    await flutterLocalNotificationsPlugin.zonedSchedule(
        0,
        title,
        body,
        tz.TZDateTime.now(tz.local).add(const Duration(seconds: 5)),
        const NotificationDetails(
            android: AndroidNotificationDetails('your channel id', 'your channel name',
                channelDescription: 'your channel description')),
        androidAllowWhileIdle: true,
        uiLocalNotificationDateInterpretation: UILocalNotificationDateInterpretation.absoluteTime);
  }


```

**Comment by MaikuB** [comment]
> What you said about TZDateTime is not the problem because I call your function _zonedScheduleNotification() almost without any changes in it.

This doesn't mean much given what I mentioned before. Furthermore, what I was referring to was if you had done your own debugging around the times being passed to the plugin. As a notification appears, it would be because it's doing so based on the conditions it was given. In this case you've could've done similar steps I had in https://github.com/MaikuB/lpn_ios_background_issue/tree/trackTimes where I reproduced the behaviour you've described. However, it is not an issue with the plugin and this all goes back to application lifecycle. In that branch I've made, I added code to track the time in shared preferences before the call to schedule the notification is made so it could be compared with the time the notification is scheduled and this would also be included in the notification's body. They are roughly 5 seconds apart as expected. However, if you look at the time that goes into shared preferences, it's actually the time the app resumes as your entire logic was put on hold when the app paused. For example, I paused the app at 8:49AM, I resumed the app at 8:50AM but what's written into shared preferences is 8:50AM and notification is appearing 5 seconds later because that is the time it was told to appear as the logic didn't run until the app had resumed

So to recap, there's no issue with the plugin here and you'll have to figure out a different solution to your problem

**Comment by misici234** [comment]
I didn't say the problem was with the plugin but I'm looking for any solution to this problem. I tested the code in Android and everything worked properly. It appears to me that iOS doesn't like to run 2 ASYNC functions together when the app goes to the background. One ASYNC is fine but 2 of them together not. 
Do you know any workaround or bandaid fix or any kind of solution for this problem?

**Comment by MaikuB** [comment]
> I didn't say the problem was with the plugin but I'm looking for any solution to this problem.

What you've said/done before would indicate otherwise, specifically by reporting a bug on this plugin, saying the notification doesn't work properly and you how it worked on an older version of the plugin. If it's not an issue with the plugin then for future please ensure you state clearly. Furthermore, this repository wouldn't be the appropriate place to ask about this as it's a generic Flutter issue. Perhaps you need to look into using the native platform's background execution APIs or use a plugin that does this so that logic can run even when the app is paused/suspend. This is something best asked elsewhere like Stack Overflow or the Flutter Slack or Discord communities but you'd need to ensure you clearly articulate the problem

**Comment by misici234** [comment]
OK you are right and I do apologize for my statement at the beginning. However, when I analysed the behaviour following the simple rule of elimination I though that Flutter SDK couldn't process web API calls in the background and that was wrong. Then I tried all other combinations and couldn't figure out the problem. Yesterday, I tried the same approach natively in iOS and it didn't work. Here is the code where pushNotification() does what it says to do.
```
    func applicationDidEnterBackground(_ application: UIApplication) {
        print("i was pushed to the background.")
        
//        perform(#selector(pushNotification), with: nil, afterDelay: 1)
//        pushNotification() // works here when commented out

        DispatchQueue.main.asyncAfter(deadline: .now() + 1.0) {
            print("Delay notification for 1 second")
            self.pushNotification() // doesn't work here
        }
    }
```
Before coming here I had tried StackOverflow but nobody helped probably because people don't have patients to read long explanations and code samples.
Anyway, thank you for your time and effort - they are greatly appreciated!
